[changed] CollapsableMixin state tracking

joemcbride opened the pull request:
* Added internal 'is expanded' tracking (see comments in mixin)
* Added basic `CollapsablePanel` example
* Added aria-expanded attributes to Panel
* Fixes expand/collapse animation for stand-alone panel
* Fixes #399, where the panel would stay a fixed size

Fixes #399

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,103 +1,133 @@
 
+/* This mixin keeps track of its own 'internal' expanded state.  This is to
+ * help enable running css animations on the collapsable component.  The
+ * expanded state /cannot/ bet set /before/ the mixin has had a chance to
```
misspelled "be" Change "bet" -> "be"

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,103 +1,133 @@
+/* This mixin keeps track of its own 'internal' expanded state.  This is to
+ * help enable running css animations on the collapsable component.  The
+ * expanded state /cannot/ bet set /before/ the mixin has had a chance to
+ * ensure that dimensions have been set on the collaspable component.  This
```
change "collaspable" -> "collapsable"

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,103 +1,133 @@
+ * help enable running css animations on the collapsable component.  The
+ * expanded state /cannot/ bet set /before/ the mixin has had a chance to
+ * ensure that dimensions have been set on the collaspable component.  This
+ * is due to how browers interact with css transitions.  If the mixin does 
```
change "browers" -> "browsers"

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,103 +1,133 @@
 var React = require('react');
 var TransitionEvents = require('./utils/TransitionEvents');
```
Can we pull this directly from React at `react/lib/ReactTransitionEvents`?

mtscout6 replied:
Or would this work: http://facebook.github.io/react/docs/animation.html#low-level-api-reacttransitiongroup

joemcbride replied:
>When children are declaratively added or removed from it (as in the example above) special lifecycle hooks are called on them

We're not adding/removing elements so I don't think we can use `ReactTransitionGroup`.

I did a diff on the core `ReactTransitionEvents` and it looks like we should be able to use that.  The note in our utils says our version is modified, but the diff says otherwise.

mtscout6 replied:
Cool!

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,103 +1,133 @@
+      this.setState({
+        collapsing:false
+      });
+    }).bind(this);
```
Man we really need to change this lib to use babel...

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,103 +1,133 @@
+       this.internalToggle();
+     }
+   }
+*/
```
I can't say that I completely understand why we can't continue to allow `expanded` to be used on the components that use this mixin. The point of this mixin is to isolate that logic away from the components that use it. If the mixin needs to retain it's own internal representation of `expanded` then use something internal such as `__collapsableMixinExpanded__`.

joemcbride replied:
>I can't say that I completely understand why we can't continue to allow expanded to be used on the components that use this mixin.

They can use their own `expanded`.  The difference is previously this mixin was relying on that `external expanded` property instead of its own.

I know this feels a little hacky, but if you look at the [bootstrap collapse](https://github.com/twbs/bootstrap/blob/master/js/collapse.js) component this is exactly how they do it.  It is just a bit less code since they're not working within a structured component framework and not keeping track of state.

If the component doesn't keep its own representation of expanded then there is no way to set `node.style['height'] = '150px'` *before* the `collapsing=true` state is set.  If that fixed height style isn't set *before* the `collapsing=true` state is set, Safari will *never* run the css animation.  The css transition is defined on the `collapsing` style.

Trying to set `node.style['height'] = '150px'` in `componentWillUpdate` doesn't work, I believe because the `collapsing` class and the dimension is set essentially at the same time.  Safari just won't run the animation.  Chrome, and I think Firefox, handle this fine just not Safari.

This wasn't an issue previously since the dimension was not being removed.  Clearing out the dimension with `node.style['height'] = ''` makes all this unfortunately necessary (at least this is the only way I could get it working).

We can change the internal expanded to so some crazy name, no problem.

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,103 +1,133 @@
+ * order in which the dimension and `expanded` information is set.
+
+ * This pattern requires the owner of this mixin to manually call the
+ * internalToggle() function to toggle the expanded/collapsed state instead
```
Having a public api method named internal feels wrong.

mtscout6 commented on docs/examples/CollapsablePanel.js:
```diff
@@ -0,0 +1,72 @@
+var CollapsablePanel = React.createClass({
```
For this example the name `Panel` feels like the Bootstrap Panel. Can we rename this to `CollapsableParagraph`?

joemcbride replied:
That was the intent  - a simplified version of the existing `Panel`.  I can change the name if you think that would be confusing.

joemcbride replied:
Simplified this a bunch.

mtscout6 replied:
Since it does not look like a `Panel` I'd prefer to rename it.

mtscout6 commented on docs/examples/CollapsablePanel.js:
```diff
@@ -0,0 +1,72 @@
+    var text = this.isExpanded() ? 'Hide' : 'Show';
+    return (
+      <div>
+        <div><Button onClick={this.onHandleToggle}>{text} Content</Button></div>
```
We can drop this div wrapper

mtscout6 commented on src/Panel.jsx:
```diff
@@ -23,22 +35,29 @@ var Panel = React.createClass({
+    } else {
+      this.handleClickEvent(e);
+      this.handleToggle();
     }
```
Following the discussion on #419 instead of an else call here we can deal with the toggle if `e.selected` is true, and leave the e.preventDefault decision to the dev handling the `onSelect` event. I'm thinking something like this:

```
handleSelect: function(e) {
  e.selected = true;

  if (this.props.onSelect) {
    this.props.onSelect(e, this.props.eventKey);
  }

  if (e.selected) {
    this.handleToggle();
  }
}
```

joemcbride replied:
`onSelect` is optional, so if its not provided wouldn't we need to call `e.prevendDefault()`?

mtscout6 replied:
If the underlying component is a button and not an anchor we should be fine.

mtscout6 commented on src/Panel.jsx:
```diff
@@ -23,22 +35,29 @@ var Panel = React.createClass({
-    this.setState({
-      expanded: !this.state.expanded
-    });
+  handleToggle: function(){
```
Thanks makes sense to be the method name in the mixin. Ideally i'd like to figure out how to have the mixin basically handle this toggling itself.

joemcbride replied:
I guess it could be possible to move the `componentDidUpdate` code to the mixin.  The `Panel` would still be responsible for setting `expanded=true` and any component that used the mixin would be stuck with using the `expanded` property.  That's probably not a huge deal though.

joemcbride replied:
I think part of the reason why I removed the `expanded` tracking from the mixin is that it started to get confusing when you should use the `internal expanded` vs. the `external expanded`.  With only one set of state tracking in the mixin then it's very clear.

joemcbride commented:
Moved the previous `componentDidMount` code into the mixin and added `_`'s to all 'internal' functions.  This simplifies the api quite a bit.  The mixin now only relies upon its owner to toggle the expanded state/prop value.

Added `aria-expanded` attributes.

joemcbride commented:
So - spent a few more hours on this (whoohoo?).  Good news is I *was* able to remove the extra state tracking.  Setting the dimension in `componentWillUpdate` is working - perhaps I was trying it in `componentWillReceiveProps` before?  I don't know - too many hours at the keyboard.  Anyway - I added a second commit so the changes could be seen.  The tests cleaned up nice.  I can squash these and swap to `[fixed]` or `[added]` after review.

mtscout6 commented:
Can you also update that example in the docs with the latest changes?

mtscout6 commented on docs/examples/CollapsableParagraph.js:
```diff
@@ -0,0 +1,37 @@
+  onHandleToggle: function(e){
+    e.preventDefault();
+    this.setState({expanded:!this.state.expanded});
+  },
```
I think this method should reside in the mixin as well. I'd rename it to `handleToggle` though since it is in response to an event and not an event itself.

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,101 +1,149 @@
-    collapsable: React.PropTypes.bool,
     defaultExpanded: React.PropTypes.bool,
     expanded: React.PropTypes.bool
   },
```
We should add documentation for both these props.

mtscout6 commented on src/CollapsableMixin.js:
```diff
@@ -1,101 +1,149 @@
-        collapsing: true
-      });
+  componentWillUpdate: function(nextProps, nextState){
+    var willExpanded = nextProps.expanded != null ? nextProps.expanded : nextState.expanded;
```
What happens when we have been getting props and the updated props is null? Should that case be false and continue not to use the internal state of this mixin? Should we care?
